### ISSUE TITLE
MCS charging interaction with ARM VCPU faults causes a kernel crash

### ISSUE THREAD

**Issue by bennoleslie** [seL4/seL4#346]
The current implementation of MCS with ARM virtualization can lead to a kernel bug.

Specifically:

1/ The handleInterruptEntry code in MCS performs an MCS budget check as the first thing it does.

2/ It is possible for an MCS budget check to requeue the current thread onto the schedule queue. At this point we have the thread both as the active thread and on a scheduling queue.

3/ On ARM with virtualization the interrupt source could be a VPPI or VGICMaintenence interrupt. Such an interrupt is treated as a fault, and an IPC is sent from the VCPU’s thread to it’s fault handler. This occurs as a a blocking synchronous IPC.

4/ After sending a blocking synchronous IPC the VCPU’s thread shall be blocked (either on the send or reply phase). However, as the MCS budget check code could have placed the thread on the scheduling queue at this point the thread is both blocked and on a scheduling queue.

Note: As far as I understand the IPC and scheduling code, threads are assume to not be on the scheduling queue at kernel entry, and therefore there is no code path to remove them from the scheduling queue if the thread is blocked during the kernel.

5/ Sometime later the scheduler will run and pick the blocked thread from the scheduling queue, at which point the kernel crashes.

Obviously this is a nasty case to trigger, and is timing sensitive. The VPPI/VGICMaint interrupt needs to come in just as the current budget is expiring.

As a work around for my immediate use case I’ve updated to the handleInterruptEntry to not perform the MCS budget check in the case where the interrupt could be a VGICMaint or VPPI interrupt. This avoids the problem and I can run without any crashes.

However, it’s not clear what the correct fix is.

I think at least some of the following should be involved:

1/ It should be invariant that on kernel exit (and entry) the current thread is not on a scheduling queue. We should assert such a variant in more places (that would have made this bug easier to find, so will likely make future bugs easier to find!).

2/ It should be invariant that prior to any code that can cause a thread to block it must not be on a scheduling queue. Examples are sendIPC, but likely others. This should also be checked more often. In fact, if the currently active thread is blocked, it should be asserted that it is not currently on the scheduled queue.

3/ The budget check on handleInterruptEntry needs to be rethought. I don’t really have a good proposal here, as I don’t understand the MCS invariants well enough. I would mention that it is odd that the code path here is slightly different to the mcsIRQ code path (or at least odd to me!).

I assume this current behaviour is not caught with formal verification as the VPPI and VGICMaintenence paths are not considered for formal verification.

If someone with more depth of understanding of MCS can provide direction I’m happy to handle to pull request side of things.

See also: https://sel4.discourse.group/t/mcs-charging-interaction-with-arm-vcpu-faults-causes-a-kernel-crash/339

**Comment by xurtis** [comment]
Based on descriptions provided elsewhere, looking at the code I think I can see what might be going wrong.

Firstly, any time that chargeBudget is called and returns `false`, we must not perform any invocation or IPC from the current thread as it may have been placed on an IPC queue (endpoint or reply) for a fault or may have been postponed and been placed in the release queue.

The non-hyp implementation of `handleInterrupt` assumes that the result of an interrupt is _never_ an IPC or invocation from the current thread, it may be a signal of a notification from the interrupt source. Thus it [calls `chargeBudget` and proceeds regardless of the result](https://github.com/seL4/seL4/blob/5b2a9250e3cdfd5caff59ee0dd380e674766102e/src/api/syscall.c#L43). This is fine for the non-hypervisor implementation, but `ARM_HYP` models some of the IRQs as faults which can produce an IPC from the currently executing thread.

When the currently executing thread ends up in the release queue or on an IPC queue and it is then enqueued for IPC again in `sendIPC` on an endpoint or reply object the following can occur:

 *  if it is in the release queue it will remain there and become rescheduled,
 *  if it was blocked on an endpoint for a fault that IPC queue will become invalid as it will still refer to the thread without the symmetric reference, or
 * if it was blocked on a reply object then that reply object will still refer to it (though there should be an assertion failure in `reply_push`)

Fundamentally, we need to reconsider how we handle the case in which `handleInterruptEntry` calls `checkBudget` and returns false. If these faults will recur if we reschedule the thread at the same point, we can simply pass the result of `checkBudget` through to `handleInterrupt` and `handleReservedIRQ` to drop any IRQs managed as faults.

**Comment by bennoleslie** [comment]
I think you description of what's happening is consistent with my understanding of what is going on.

I do not think that the faults will recur is the thread is rescheduled at the same point. If they are not handled they will be lost. (Is my understanding, someone with more detailed understanding of ARM virtualization may have a better understanding.)

I think we need to check for these interrupts before calling checkBudget. (Unless there is a way checkBudget can be moved after the interrupt handling in general. It's not clear to me why it couldn't, but I'm not up to speed on MCS sufficiently to be confident in that).

**Comment by kent-mcleod** [comment]
> I do not think that the faults will recur is the thread is rescheduled at the same point.

Specifically for VGIC and VPPI this may not be an issue as both IRQ sources are level triggered, so when the VCPU thread is rescheduled the IRQ conditions will be raised again:
- VGIC IRQs are specified as level triggered in the GICv2 spec and I'd expect that they'd stay that way in future revisions as it's necessary to handle situations where the EL2 hypervisor already chose to reschedule the guest thread before it saw the VGIC IRQ.  In these cases it's supposed to be safe to ignore the IRQ and expect that it will be regenerated once the guest is switched back to.  I'd expect that this state is tracked as part of the GICH (GIC virtual interface control registers) that are context switched by the EL2 hypervisor.
- VPPI irqs currently only consist of the generic timer Vtimer interrupt.  This is also a level triggered IRQ and if it is ignored by the kernel it also won't be lost.

So currently I think dropping the IRQ if it is a VGIC or VPPI should work for now, but I agree that we should have a way to resolve this once we encounter an IRQ-fault scenario that isn't level triggered.  In that case I think we'll need to do something like save a continuation in the thread object for resuming the IRQ-fault handling once the thread is rescheduled.

**Comment by kent-mcleod** [comment]
It's not possible to ignore the checkBudget if the IRQ isn't a kernel timer interrupt as there may not be enough time left for that SC to wake up and reenter the kernel.

**Comment by bennoleslie** [comment]
OK. So this change is going to be pretty invasive as I understand thing.

1/ Update `handleInterrupt` so that it takes a `bool hasBudget` or some such thing.
2/ Update `handleReservedIRQ` similarly.
3/ Change implement of `handleReservedIRQ` on ARM  so that `VGICMaintenance` and `VPPIEvent` are only called in the case where `hasBudget` is true.
4/ Update `handleInterruptEntry` so that the return value `checkBudget` is passed to `handleInterrupt`
5/ `handleInterrupt` is also called on other paths within `syscall.c`. Each of these would also need to passed the `hasBudget` appropriately. I assume this would required `mcsIRQ` to also return if the current thread has budget.

In doing all the above ensure it is conditional on `CONFIG_KERNEL_MCS`. This seems like it will be somewhat verbose, but not too bad.

**Comment by kent-mcleod** [comment]
I think so.  There's likely a way to calculate what hasBudget would be from the global state and it would allow implementing this without having to change all the function signatures.

**Comment by bennoleslie** [comment]
Well, I can we could just skip it is the current thread is enqueued. That's really the exact thing that causes the problem and is a fast check. For some reason is feels wrong, but as far as I can tell there would be no other case where the active thread is enqueued. Thoughts?

**Comment by bennoleslie** [comment]
Pull request: https://github.com/seL4/seL4/pull/352

**Comment by xurtis** [comment]
@bennoleslie is this now fixed by #352? If so, can we close this issue?

**Comment by kent-mcleod** [comment]
Assuming this was resolved by #352.
